scripts/pythondistdeps.py

```diff
@@ -381,26 +415,39 @@ def get_marker_env(dist, extra):
                         deps.insert(0, Requirement('setuptools'))
                 # add requires/recommends based on egg/dist metadata
                 for dep in deps:
-                    if normalized_names_require_pep503:
-                        dep_normalized_name = dep.normalized_name
-                    else:
-                        dep_normalized_name = dep.legacy_normalized_name
-
-                    if args.legacy:
-                        name = 'pythonegg({})({})'.format(pyver_major, dep.legacy_normalized_name)
-                    else:
-                        if args.majorver_only:
-                            name = 'python{}dist({})'.format(pyver_major, dep_normalized_name)
+                    # Even if we're requiring `foo[bar]`, also require `foo`
+                    # to be safe, and to make it discoverable through
+                    # `repoquery --whatrequires`
+                    extras_suffixes = [""]
+                    if args.require_extras_subpackages and dep.extras:
+                        # A dependency can have more than one extras,
+                        # i.e. foo[bar,baz], so let's go through all of them
+                        extras_suffixes += [f"[{e}]" for e in dep.extras]
+
+                    for extras_suffix in extras_suffixes:
+                        if normalized_names_require_pep503:
+                            dep_normalized_name = dep.normalized_name
                         else:
-                            name = 'python{}dist({})'.format(dist.py_version, dep_normalized_name)
-                    if dep.marker and not args.recommends:
-                        if not dep.marker.evaluate(get_marker_env(dist, '')):
-                            continue
-                    if name not in py_deps:
-                        py_deps[name] = []
-                    for spec in dep.specifier:
-                        if (spec.operator, spec.version) not in py_deps[name]:
-                            py_deps[name].append((spec.operator, spec.version))
+                            dep_normalized_name = dep.legacy_normalized_name
+
+                        if args.legacy:
+                            name = 'pythonegg({})({})'.format(pyver_major, dep.legacy_normalized_name)
+                        else:
+                            if args.majorver_only:
+                                name = 'python{}dist({}{})'.format(pyver_major, dep_normalized_name, extras_suffix)
+                            else:
+                                name = 'python{}dist({}{})'.format(dist.py_version, dep_normalized_name, extras_suffix)
+
+                        if dep.marker and not args.recommends and not extras_subpackage:
+                            if not dep.marker.evaluate(get_marker_env(dist, '')):
+                                continue
```

Comment:
Wrapping my head around this merge is hard :/

Reply:
Any explanation is appreciated.

Reply:
This is not a new piece of code, I only indented it and added the `not extras_subpackage`.

This is how I understand the piece of code:

This part of code is run for all `deps`, which are all requirements of the given package. When processing a non-extras Python package, the list of requirements includes all its direct requirements, but also the requires of all its extras packages. The information whether it's a direct requires or a requires of one of the extras packages is stored in `dep.marker`.

So, this piece of code takes each requires (`dep`) and evaluates if it's a direct dependency (e.g. a requires for an extras package `''`). But if the `args.recommends` is enabled this is skipped, because all the dependencies of `extras` will be recommended. And of course it should be skipped when we're processing an extras subpackage and its requires.